File: 05-binary_search/00981-Time_Based_Key-Value_Store/00981-time_based_key-value_store.java
// 981: Time Based Key-Value Store
// https://leetcode.com/problems/time-based-key-value-store/


import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;


class Solution {
    static class Data {
        String val;
        int time;
        Data(String val, int time) {
            this.val = val;
            this.time = time;
        }
    }

    // SOLUTION
    static class TimeMap {
        HashMap<String, List<Data>> tm;

        public TimeMap() {
            tm = new HashMap<>();
        }

        public void set(String key, String value, int timestamp) {
            if (!tm.containsKey(key)) tm.put(key, new ArrayList<Data>());
            tm.get(key).add(new Data(value, timestamp));
        }
        
        public String get(String key, int timestamp) {
            if (!tm.containsKey(key)) return "";
            List<Data> list = tm.get(key);
            
            int l = 0;
            int h = list.size() - 1;

            while (l < h) {
                int m = l + (h-l+1) / 2;
                
                if (list.get(m).time <= timestamp) l = m;
                else h = m - 1;
            }

            return list.get(l).time <= timestamp ? list.get(l).val : "";
        }
    }

    public static void main(String[] args) {
        // OPERATIONS
        TimeMap o = new TimeMap();
        o.set("foo", "bar", 1);
        System.out.println(o.get("foo", 1));
        System.out.println(o.get("foo", 3));
        o.set("foo", "bar2", 4);
        System.out.println(o.get("foo", 4));
        System.out.println(o.get("foo", 5));
    }
}

File: 05-binary_search/00981-Time_Based_Key-Value_Store/00981-time_based_key-value_store.py
# 981: Time Based Key-Value Store
# https://leetcode.com/problems/time-based-key-value-store/


import collections

class TimeMap(object):
    def __init__(self):
        self.tm = collections.defaultdict(list)

    def set(self, key, value, timestamp):
        self.tm[key].append([timestamp, value])

    def get(self, key, timestamp):
        tlist = self.tm[key]

        l = 0
        h = len(tlist)
        
        while l < h:
            m = (l + h) // 2
            
            if tlist[m][0] <= timestamp:
                l = m + 1
            elif tlist[m][0] > timestamp:
                h = m
        
        return "" if h == 0 else tlist[h - 1][1]



if __name__ == "__main__":
    # OPERATIONS
    o = TimeMap()
    o.set("foo", "bar", 1);
    print(o.get("foo", 1));
    print(o.get("foo", 3));
    o.set("foo", "bar2", 4);
    print(o.get("foo", 4));
    print(o.get("foo", 5));
